controllers/utils/utils.go

```diff
@@ -0,0 +1,128 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	capiv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	controlplanev1 "sigs.k8s.io/cluster-api/controlplane/kubeadm/api/v1beta1"
+	"sigs.k8s.io/cluster-api/util"
+	clientPkg "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GetMachineSet attempts to fetch a MachineSet from CAPI machine owner reference.
+func GetMachineSetFromCAPIMachine(
+	ctx context.Context,
+	client clientPkg.Client,
+	capiMachine *capiv1.Machine,
+) (*capiv1.MachineSet, error) {
+
+	ref := GetManagementOwnerRef(capiMachine)
+	gv, err := schema.ParseGroupVersion(ref.APIVersion)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if gv.Group == capiv1.GroupVersion.Group {
+		key := clientPkg.ObjectKey{
+			Namespace: capiMachine.Namespace,
+			Name:      ref.Name,
+		}
+
+		machineSet := &capiv1.MachineSet{}
+		if err := client.Get(ctx, key, machineSet); err != nil {
+			return nil, err
+		}
+
+		return machineSet, nil
+	}
+	return nil, nil
+}
+
+// GetKubeadmControlPlaneFromCAPIMachine attempts to fetch a KubeadmControlPlane from a CAPI machine owner reference.
+func GetKubeadmControlPlaneFromCAPIMachine(
+	ctx context.Context,
+	client clientPkg.Client,
+	capiMachine *capiv1.Machine,
+) (*controlplanev1.KubeadmControlPlane, error) {
+
+	ref := GetManagementOwnerRef(capiMachine)
+	gv, err := schema.ParseGroupVersion(ref.APIVersion)
+	if err != nil {
+		return nil, errors.WithStack(err)
+	}
+	if gv.Group == controlplanev1.GroupVersion.Group {
+		key := clientPkg.ObjectKey{
+			Namespace: capiMachine.Namespace,
+			Name:      ref.Name,
+		}
+
+		controlPlane := &controlplanev1.KubeadmControlPlane{}
+		if err := client.Get(ctx, key, controlPlane); err != nil {
+			return nil, err
+		}
+
+		return controlPlane, nil
+	}
+	return nil, nil
+}
+
+// IsOwnerDeleted returns a boolean if the owner of the CAPI machine has been deleted.
+func IsOwnerDeleted(ctx context.Context, client clientPkg.Client, capiMachine *capiv1.Machine) (bool, error) {
+	if util.IsControlPlaneMachine(capiMachine) {
+		// The controlplane sticks around after deletion pending the deletion of its machiens.
+		// As such, need to check the deletion timestamp thereof.
+		if cp, err := GetKubeadmControlPlaneFromCAPIMachine(ctx, client, capiMachine); cp != nil && cp.DeletionTimestamp == nil {
+			return false, nil
+		} else if err != nil && !strings.Contains(err.Error(), "not found") {
+			return false, err
+		}
+	} else {
+		// The machienset is deleted immediately, regardless of machine ownership.
+		// It is sufficient to check for its existence.
+		if ms, err := GetMachineSetFromCAPIMachine(ctx, client, capiMachine); ms != nil {
+			return false, nil
+		} else if err != nil && !strings.Contains(err.Error(), "not found") {
+			return false, err
+		}
+	}
+	return true, nil
+}
+
+// fetchRef simply searches a list of OwnerReference objects for a given kind.
+func fetchRef(reflist []meta.OwnerReference, kind string) *meta.OwnerReference {
```

Comment:
nit:fetchOwnerRef*? And/or capitalize the L in refList for proper camel casing

Reply:
Did both.